verification/provenance.go

```diff
@@ -34,6 +34,89 @@ func provenanceFromEnv(env *dsselib.Envelope) (prov *intoto.ProvenanceStatement,
 	return
 }
 
+// Verify Builder ID in provenance statement.
+func verifyBuilderID(prov *intoto.ProvenanceStatement, builderID string) error {
+	id, err := sourceFromURI(prov.Predicate.Builder.ID)
+	if err != nil {
+		return err
+	}
+	if !strings.EqualFold(id, builderID) {
+		return fmt.Errorf("%w: expected '%s' in builder.id, got '%s'", ErrorMismatchBuilderID,
+			builderID, id)
+	}
+	return nil
+}
+
+func asURI(s string) string {
+	source := s
+	if !strings.HasPrefix(source, "https://") &&
+		!strings.HasPrefix(source, "git+") {
+		source = "git+https://" + source
+	}
+	if !strings.HasPrefix(source, "git+") {
+		source = "git+" + source
+	}
+
+	return source
+}
+
+// Verify source URI in provenance statement.
+func verifySourceURI(prov *intoto.ProvenanceStatement, expectedSourceURI string) error {
+	source := asURI(expectedSourceURI)
+
+	// We expect github.com URIs only.
+	if !strings.HasPrefix(source, "git+https://github.com/") {
+		return fmt.Errorf("%w: expected source github.com repository '%s'", ErrorMalformedURI,
+			source)
+	}
+
+	// Verify source from ConfigSource field.
+	configURI, err := sourceFromURI(prov.Predicate.Invocation.ConfigSource.URI)
+	if err != nil {
+		return err
+	}
+	if !strings.EqualFold(configURI, source) {
```

Comment:
Again, I suspect that unicode folding isn't how we likely want to do URI comparisons.

Reply:
Good catch. I'll change that. Thank you!